Allow the serializer to return sparse fieldsets

Erol opened the pull request:
I also need sparse fieldset support from Fast JSON API. After reading through and seeing that there's no recent activity on #57, #63, and #254, I'd like to take a stab at this.

A client should be able to request only a subset of attributes and relationships using the `fields` option. Here's an example using the movie context from the test suite:

```ruby
options = {
  fields: {
    movie: %i[name actors advertising_campaign],
    actor: %i[name agency]
  },
  include: %i[actors advertising_campaign]
}

MovieSerializer.new(movie, options).serialized_json
```

Note that this would respect both attributes and relationships when the fields have been explicitly given. The resulting `movie` would only return `name` in its `attributes`, and only `actors`, `advertising_campaign` would be retured in its `relationships`.

Fields will also be filtered on included documents. All `actor` will only return `name` and `agency` on their `attributes` and `relationships` respectively.

Records which do not have explicit fields will return all its attributes and relationships, as shown by `advertising_campaign`.

Here's the full JSON:

```json
{
  "data":{
    "id":"232",
    "type":"movie",
    "attributes":{
      "name":"test movie"
    },
    "relationships":{
      "actors":{
        "data":[
          {
            "id":"1",
            "type":"actor"
          },
          {
            "id":"2",
            "type":"actor"
          },
          {
            "id":"3",
            "type":"actor"
          }
        ]
      },
      "advertising_campaign":{
        "data":{
          "id":"1",
          "type":"advertising_campaign"
        }
      }
    }
  },
  "included":[
    {
      "id":"1",
      "type":"actor",
      "attributes":{
        "name":"Test 1"
      },
      "relationships":{
        "agency":{
          "data":{
            "id":"0",
            "type":"agency"
          }
        }
      }
    },
    {
      "id":"2",
      "type":"actor",
      "attributes":{
        "name":"Test 2"
      },
      "relationships":{
        "agency":{
          "data":{
            "id":"1",
            "type":"agency"
          }
        }
      }
    },
    {
      "id":"3",
      "type":"actor",
      "attributes":{
        "name":"Test 3"
      },
      "relationships":{
        "agency":{
          "data":{
            "id":"2",
            "type":"agency"
          }
        }
      }
    },
    {
      "id":"1",
      "type":"advertising_campaign",
      "attributes":{
        "id":1,
        "name":"Movie test movie is incredible!!"
      },
      "relationships":{
        "movie":{
          "data":{
            "id":"232",
            "type":"movie"
          }
        }
      }
    }
  ]
}
```

Erol commented:
Hope you can have a look at this as well @shishirmk

shishirmk commented:
will take a look at this and get back to you in a day..

Erol commented:
Thanks. Let me know if there's anything else I need to consider

shishirmk commented on lib/fast_jsonapi/fieldset.rb:
```diff
@@ -0,0 +1,7 @@
+module FastJsonapi
+  class Fieldset
+    def initialize(fields)
+      @fields = fields
```
could we run define and run deep_symbolize method here? I think it makes sense for it to be transparent to the rest of the code.

Erol replied:
That's a remnant of a previous attempt and no longer needed. I'll delete it and remove the reference

shishirmk commented on lib/fast_jsonapi/object_serializer.rb:
```diff
@@ -41,8 +42,8 @@ def hash_for_one_record
 
-      serializable_hash[:data] = self.class.record_hash(@resource, @params)
-      serializable_hash[:included] = self.class.get_included_records(@resource, @includes, @known_included_objects, @params) if @includes.present?
+      serializable_hash[:data] = self.class.record_hash(@resource, @fieldsets[self.class.reflected_record_type.to_sym], @params)
```
`reflected_record_type` is the meaningful default it can be overridden in `set_type` so it probably makes sense here to use `self.record_type`

Erol replied:
👍

shishirmk commented on lib/fast_jsonapi/object_serializer.rb:
```diff
@@ -51,9 +52,10 @@ def hash_for_collection
 
       data = []
       included = []
+      fieldset = @fieldsets[self.class.reflected_record_type.to_sym]
```
`reflected_record_type` is the meaningful default it can be overridden in `set_type` so it probably makes sense here to use `self.record_type`

Erol replied:
👍

Erol replied:
@shishirmk Seems like `record_type` is not being set for inherited serializers. Got this when I introduced a spec for it:

```
  1) FastJsonapi::ObjectSerializer sets the correct record type
     Failure/Error: expect(EmployeeSerializer.record_type).to eq 'employee'

       expected: "employee"
            got: nil

       (compared using eql?)
     # ./spec/lib/object_serializer_inheritance_spec.rb:99:in `block (2 levels) in <top (required)>'
```

I'll write a fix for that as well.

Erol replied:
@shishirmk Looks like `record_type` is not being set for inherited serializers. I added a spec for it and got this:

```
  1) FastJsonapi::ObjectSerializer sets the correct record type
     Failure/Error: expect(EmployeeSerializer.record_type).to eq 'employee'

       expected: "employee"
            got: nil

       (compared using ==)
     # ./spec/lib/object_serializer_inheritance_spec.rb:99:in `block (2 levels) in <top (required)>'
```

Will write a fix for that as well.

shishirmk replied:
Yup go ahead and fix that as well.

shishirmk commented on lib/fast_jsonapi/object_serializer.rb:
```diff
@@ -104,6 +120,7 @@ def inherited(subclass)
         subclass.race_condition_ttl = race_condition_ttl
         subclass.data_links = data_links
         subclass.cached = cached
+        subclass.fieldset = fieldset.dup if fieldset.present?
```
fieldsets needs to be passed in each time `serializable_hash` is invoked. I don't think it makes sense to inherit this when subclassing.

Erol replied:
That's another remnant I failed to cleanup. Should be gone after the next commit.

shishirmk commented on lib/fast_jsonapi/serialization_core.rb:
```diff
@@ -40,36 +41,39 @@ def links_hash(record, params = {})
-        attributes_to_serialize.each_with_object({}) do |(_k, attribute), hash|
+      def attributes_hash(record, fieldset = nil, params = {})
+        attributes = attributes_to_serialize
+        attributes = attributes.slice(*fieldset) if fieldset.present?
```
running this slice for each record in an array is probably unnecessary. It would be nice if we can avoid that. Not needed for merging in.

Erol replied:
Yeah that's the ideal. I'll open another PR as an improvement.

shishirmk commented on lib/fast_jsonapi/serialization_core.rb:
```diff
@@ -40,36 +41,39 @@ def links_hash(record, params = {})
-      def relationships_hash(record, relationships = nil, params = {})
+      def relationships_hash(record, relationships = nil, fieldset = nil, params = {})
         relationships = relationships_to_serialize if relationships.nil?
+        relationships = relationships.slice(*fieldset) if fieldset.present?
```
running this slice for each record in an array is probably unnecessary. It would be nice if we can avoid that. Not needed for merging in.

shishirmk commented:
@Erol This looks really good. Thank you for the work you have put in. I have added a few comments please address them. Add how to use this feature to the read me. Then we can merge it in and I will release it in the next version by the end of the week.

Erol commented:
@shishirmk Addressed your feedback with the commits above. I've also added instructions on the README.

Also added a fix to set the `record_type` on inherited serializers here: https://github.com/Netflix/fast_jsonapi/commit/e7dc8b4b31857c39295bdd5310e00d6a490ba6ce. That might partially resolve #270.

g13ydson commented:
@Erol @shishirmk I have a problem. I have created the following serializers:
```ruby
class UserSerializer
  include FastJsonapi::ObjectSerializer
  attributes :name_badge: 
  has_many :occupations
end
```

```ruby
class OccupationSerializer
  include FastJsonapi::ObjectSerializer
  attributes :type
end
```

When I try:
`UserSerializer.new(@user, {include: [:occupations], fields: {user: [:name]}})
`

I get as return the JSON below, notice that the relationships field is empty, should not come filled?

```json
{
    "data": {
        "id": "1",
        "type": "user",
        "attributes": {
            "name": "Gleydson"
        },
        "relationships": {}
    },
    "included": [
        {
            "id": "1",
            "type": "occupation",
            "attributes": {
                "type": "Teacher"
            }
        }
    ]
}
```

shishirmk commented:
I dont think relationships should be filled in according to the spec. Here is the link http://jsonapi.org/format/#fetching-sparse-fieldsets
> If a client requests a restricted set of fields for a given resource type, an endpoint MUST NOT include additional fields in resource objects of that type in its response.
